compiler/src/dotty/tools/dotc/transform/PickleQuotes.scala

```diff
@@ -458,105 +170,215 @@ class PickleQuotes extends MacroTransform {
+        case Constants.StringTag => liftedValue(lit, defn.ToExprModule_StringToExpr)
+      }
     }
 
-    /** Type used for the hole that will replace this splice */
-    def getHoleType(body: tpd.Tree, splice: tpd.Tree)(using Context): Type =
-      // For most expressions the splice.tpe but there are some types that are lost by lifting
-      // that can be recoverd from the original tree. Currently the cases are:
-      //  * Method types: the splice represents a method reference
-      map.get(body.symbol).map(_.tpe.widen).getOrElse(splice.tpe)
+    /** Encode quote using QuoteUnpickler.{unpickleExpr, unpickleType}
+      *
+      *  Generate the code
+      *  ```scala
+      *    quotes => quotes.asInstanceOf[QuoteUnpickler].<unpickleExpr|unpickleType>[<type>](
+      *      <pickledQuote>,
+      *      <typeHole>,
+      *      <termHole>,
+      *    )
+      *  ```
+      *  this closure is always applied directly to the actual context and the BetaReduce phase removes it.
+      */
+    def pickleAsTasty() = {
+      val pickleQuote = PickledQuotes.pickleQuote(body)
+      val pickledQuoteStrings = pickleQuote match
+        case x :: Nil => Literal(Constant(x))
+        case xs => tpd.mkList(xs.map(x => Literal(Constant(x))), TypeTree(defn.StringType))
+
+      // TODO split holes earlier into types and terms. This all holes in each category can have consecutive indices
+      val (typeSplices, termSplices) = contents.zipWithIndex.partition {
+        case (splice, _) => splice.tpe.derivesFrom(defn.QuotedTypeClass)
+      }
+
+      // This and all closures in typeSplices are removed by the BetaReduce phase
+      val typeHoles =
+        if typeSplices.isEmpty then Literal(Constant(null)) // keep pickled quote without contents as small as possible
+        else
+          Lambda(
+            MethodType(
+              List("idx", "contents").map(name => UniqueName.fresh(name.toTermName).toTermName),
+              List(defn.IntType, defn.SeqType.appliedTo(defn.AnyType)),
+              defn.QuotedTypeClass.typeRef.appliedTo(WildcardType)),
+            args => {
+              val cases = typeSplices.map { case (splice, idx) =>
+                CaseDef(Literal(Constant(idx)), EmptyTree, splice)
+              }
+              cases match
+                case CaseDef(_, _, rhs) :: Nil => rhs
+                case _ => Match(args(0).annotated(New(ref(defn.UncheckedAnnot.typeRef))), cases)
+            }
+          )
 
-    def isLiftedSymbol(sym: Symbol)(using Context): Boolean = map.contains(sym)
+      // This and all closures in termSplices are removed by the BetaReduce phase
+      val termHoles =
+        if termSplices.isEmpty then Literal(Constant(null)) // keep pickled quote without contents as small as possible
+        else
+          Lambda(
+            MethodType(
+              List("idx", "contents", "quotes").map(name => UniqueName.fresh(name.toTermName).toTermName),
```

Comment:
These names can be added to StdNames to avoid the toTermName.

Reply:
Added